### PR TITLE
Add firstOrFail, findOrFail support

### PR DESCRIPTION
**Pull request by zoul0813**
This change is [Reviewable](https://reviewable.io/reviews/baopham/laravel-dynamodb/61)

### CODE REVIEW

File: src/DynamoDbQueryBuilder.php [hunk 1]
```diff
@@ -154,13 +155,34 @@ public function find($id, array $columns = [])
         return $model;
     }
 
+    public function findOrFail($id, array $columns = [])
+    {
+        $result = $this->find($id, $columns);
+        
+        if (is_array($id) && count($result) == count(array_unique($id))) {
+            return $result;
+        } elseif (! is_null($result)) {
```
Comment (baopham) [inline src/DynamoDbQueryBuilder.php]: Let's copy Eloquent's version over as is 😄

Comment (zoul0813) [inline src/DynamoDbQueryBuilder.php]: Not sure I follow - ?

Comment (baopham) [inline src/DynamoDbQueryBuilder.php]: Oh, I meant to say let's just copy over this chunk https://github.com/laravel/framework/blob/1742bc9a378162c3b57f1bc799e91f2ab0802aef/src/Illuminate/Database/Eloquent/Builder.php#L300-L315 as is. Your version is a slightly modified version of it.

Comment (zoul0813) [inline src/DynamoDbQueryBuilder.php]: Yeah, I modified it to make Nitpick happy ... I just pushed an updated version that is a verbatim "copy/paste" of Eloquent (which was what I had originally for my local tests, until Nitpick got nitpicky ...)

**Comment by zoul0813** [comment]
Working on tests now

**Comment by zoul0813** [comment]
Having trouble with the first test ... not sure why it's throwing this DynamoDbException `{"__type":"com.amazon.coral.validate#ValidationException","message":"Invalid attribute value type"}` exception

```php
    public function testFindOrFailRecord() 
    {
        $seed = $this->seed();
        $seedId = array_get($seed, 'id.S');
        $this->expectException(ModelNotFoundException::class);
        $item = $this->testModel->findOrFail('expected-to-fail');
    }
```

```
There were 2 failures:

1) BaoPham\DynamoDb\Tests\DynamoDbCompositeModelTest::testFindOrFailRecord
Failed asserting that exception of type "Aws\DynamoDb\Exception\DynamoDbException" matches expected exception "Illuminate\Database\Eloquent\ModelNotFoundException". Message was: "Error executing "GetItem" on "http://localhost:3000"; AWS HTTP error: Client error: `POST http://localhost:3000` resulted in a `400 Bad Request` response:
{"__type":"com.amazon.coral.validate#ValidationException","message":"Invalid attribute value type"}
 ValidationException (client): Invalid attribute value type - {"__type":"com.amazon.coral.validate#ValidationException","message":"Invalid attribute value type"}" at
.../laravel-dynamodb/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113
.../laravel-dynamodb/vendor/guzzlehttp/guzzle/src/Middleware.php:65
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:203
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:156
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/TaskQueue.php:47
.../laravel-dynamodb/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:96
.../laravel-dynamodb/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:123
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:246
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:223
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:267
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:225
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:267
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:225
.../laravel-dynamodb/vendor/guzzlehttp/promises/src/Promise.php:62
.../laravel-dynamodb/vendor/aws/aws-sdk-php/src/AwsClientTrait.php:59
.../laravel-dynamodb/vendor/aws/aws-sdk-php/src/AwsClientTrait.php:78
.../laravel-dynamodb/src/DynamoDbQueryBuilder.php:137
.../laravel-dynamodb/vendor/illuminate/database/Eloquent/Model.php:1368
.../laravel-dynamodb/tests/DynamoDbModelTest.php:44
```

**Comment by zoul0813** [comment]
Having an issue with the Composite Test ... 

```
> phpunit
PHPUnit 5.7.21 by Sebastian Bergmann and contributors.

.....E..........................................................  64 / 64 (100%)

Time: 1.12 seconds, Memory: 36.00MB

There was 1 error:

1) BaoPham\DynamoDb\Tests\DynamoDbCompositeModelTest::testFindOrFailRecordPass
Illuminate\Database\Eloquent\ModelNotFoundException: No query results for model [BaoPham\DynamoDb\Tests\CompositeTestModel] id1, RP2LIQJViHTDlxNUgrD3y2EOO125XO66nBEp

/Users/dhiggins/Sites/equine/tmp/laravel-dynamodb/src/DynamoDbQueryBuilder.php:170
/Users/dhiggins/Sites/equine/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Model.php:1368
/Users/dhiggins/Sites/equine/tmp/laravel-dynamodb/tests/DynamoDbCompositeModelTest.php:64

ERRORS!
Tests: 64, Assertions: 136, Errors: 1.
Script phpunit handling the test event returned with error code 2
```

Not quite sure why it's saying it can't find anything ... ran out of time for the day to work on this, so I committed what I have so far.  If you can help me figure out what's wrong with the composite test ... that would be great.

File: src/DynamoDbQueryBuilder.php [hunk 2]
```diff
@@ -154,13 +155,36 @@ public function find($id, array $columns = [])
...
+        if (is_array($id)) {
+            if (count($result) == count(array_unique($id))) {
+                return $result;
+            }
+        } elseif (! is_null($result)) {
+            return $result;
+        }
+
+        throw (new ModelNotFoundException)->setModel(
+            get_class($this->model), $id
```
Comment (zoul0813) [inline src/DynamoDbQueryBuilder.php]: ignoring - direct copy/paste from Eloquent

**Comment by zoul0813** [comment]
Ok, tracked the problem down to the composite keys ... findOrFail in Eloquent is expecting that if $id is an array, each item in the array is an ID to find ... where as DynamoDbModel is treating it as a composite key lookup.  May need to rework the way the composite keys are passed to 'find' and 'findOrFail' ... an array of arrays perhaps? As is, DynamoDbModel doesn't fully support "find" per the Eloquent documentation.

Per Eloquent, 

```
You may also call the find method with an array of primary keys, which will return a collection of the matching records:

$flights = App\Flight::find([1, 2, 3]);
```

Should I submit a new issue to add support for this, and just hack the "findOrFail" to work with "find"  the way it is now?

**Comment by zoul0813** [comment]
Added the issue, adjusted findOrFail to ignore the array of ID's to support composite keys "as is" in DynamoDbQueryBuilder.

**Comment by baopham** [comment]
To make this work elegently we need to refactor a bit the composite key work. For now, I would say we check if the model has composite key then we treat it the param as one ID.

**Comment by baopham** [comment]
Any news?

In short, I think we can do this:

```
$composite = !empty($this->compositeKey);
$isMultiple = $composite ? is_array(array_first($id)) : is_array($id);

// Proceed with the usual logic
```

**Comment by zoul0813** [comment]
hrm... maybe.  I'll be on vacation for the next few days, returning next thursday or friday ... do you mind adding that in yourself?

**Comment by baopham** [comment]
I don't think I have push access to your repo. No rush though. I'm also occupied with other stuff. 

Enjoy your vacation 🎉

**Comment by zoul0813** [comment]
Just added you as a collaborator on my fork... in the event you find free time before I do

**Comment by zoul0813** [comment]
Just took a look at this, and I'm not entirely sure I follow how the composite key structure is working ... so I'm not sure how to setup find/findOrFail to work with either an array of ID's or a composite key.  The problem seems to be in calling "setId" in the ->find() method ... 

I managed to get the Non-Composite/Composite tests to pass, but when I tried to add a second test to the non-composite to call find/findOrFail with multiple ID's it is having issues, I believe because `find()` is setup to call `setId($id)` which is then used to help build the query to DynamoDb to get a single item ... so I don't believe this is going to be a "quick fix".

I've pushed up my initial tweaks to this ... also included a fix for the ISO8601 (using DateTime::ATOM instead of DateTime::ISO8601 per the PHP Docs)

**Comment by baopham** [comment]
No worries. If you don't mind, I'll take a look this weekend. Hope this is not block you.

**Comment by zoul0813** [comment]
not blocking me on my end, I have my project setup to use my forks develop branch which is being kept in sync with my pull requests.  Whenever you commit a pull request, I just re-sync my forks develop branch to account for anything you might change or add.

**Comment by baopham** [comment]
Got some idea. I'll work on it during the week.

**Pull request merged**